plugins/inputs/firehose/firehose.go

```diff
@@ -0,0 +1,413 @@
+	select {
+	case <-f.close:
+		res.WriteHeader(http.StatusGone)
+		return
+	default:
+	}
+
+	if !choice.Contains(req.URL.Path, f.Paths) {
+		res.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	r := &firehoseRequest{req: req}
+
+	if f.AccessKey != "" {
+		if ok := r.authenticate(f.AccessKey); !ok {
+			f.Log.Error(formatLog(req.Header.Get(requestIDHeader), "Unauthorized"))
+			err := r.sendResponse(res)
+			if err != nil {
+				f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+			}
+			return
+		}
+	}
+
+	if err := r.validate(); err != nil {
+		f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+		err := r.sendResponse(res)
+		if err != nil {
+			f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+		}
+		return
+	}
+
+	decodedBytesData, ok := r.decodeData()
+	if !ok {
+		f.Log.Error(formatLog(r.body.RequestID, "Failed to base64 decode record data"))
+		err := r.sendResponse(res)
+		if err != nil {
+			f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+		}
+		return
+	}
+
+	metrics := make([]telegraf.Metric, 0)
+	for _, bytes := range decodedBytesData {
+		m, err := f.Parse(bytes)
+		if err != nil {
+			f.Log.Error(formatLog(r.body.RequestID, "Unable to parse data"))
+			r.responseStatusCode = http.StatusBadRequest
+			err := r.sendResponse(res)
+			if err != nil {
+				f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+			}
+			return
+		}
+		metrics = append(metrics, m...)
+	}
+
+	if len(metrics) == 0 {
+		once.Do(func() {
+			f.Log.Info(formatLog(r.body.RequestID, internal.NoMetricsCreatedMsg))
+		})
+	}
+
+	commonAttributesHeaderValue := req.Header.Get(commonAttributesHeader)
+	if len(commonAttributesHeaderValue) != 0 && len(f.ParameterTags) != 0 {
+		parameters := make(map[string]interface{})
+		err := json.Unmarshal([]byte(commonAttributesHeaderValue), &parameters)
+		if err != nil {
+			f.Log.Warn(formatLog(r.body.RequestID, commonAttributesHeader+" header's value is not a valid json"))
+		}
+
+		parameters, ok := parameters["commonAttributes"].(map[string]interface{})
+		if !ok {
+			f.Log.Warn(formatLog(r.body.RequestID, "Invalid value for header "+commonAttributesHeader))
+		}
+
+		for _, parameter := range f.ParameterTags {
+			if value, ok := parameters[parameter]; ok {
+				for _, m := range metrics {
+					m.AddTag(parameter, value.(string))
+				}
+			}
+		}
+	}
+
+	for _, m := range metrics {
+		if f.PathTag {
+			m.AddTag(pathTag, req.URL.Path)
+		}
+		f.acc.AddMetric(m)
+	}
+
+	r.responseStatusCode = http.StatusOK
+	err := r.sendResponse(res)
+	if err != nil {
+		f.Log.Error(formatLog(req.Header.Get(requestIDHeader), err.Error()))
+	}
+}
+
+func formatLog(requestID, message string) string {
+	return fmt.Sprintf(logFormat, requestID, message)
+}
+
+func init() {
+	inputs.Add("firehose", func() telegraf.Input {
+		return &Firehose{
+			ServiceAddress: ":8080",
+			TimeFunc:       time.Now,
+			Paths:          []string{"/telegraf"},
```

Comment:
Please set the default values of `ServiceAddress` and `Paths` in `Init()` to reuse it in tests.

Reply:
Makes sense. Btw if it's only for tests can't we do it the way it's done [here](https://github.com/influxdata/telegraf/blob/a16763497b89a65a536892e4ac8431506e3afc26/plugins/inputs/http_listener_v2/http_listener_v2_test.go#L52)? Just a thought

Reply:
TBH, I hate these types of indirections. Imagine a test fails and you quickly want to look at what's going on. You immediately need to jump around in code to see how a simple test case is set up... :-( So please avoid this as possible!

Reply:
done